packages/context/src/context-view.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  ContextObserver,
+  ContextEventType,
+  Subscription,
+} from './context-observer';
+import {Getter} from './inject';
+import {ResolutionSession} from './resolution-session';
+import {resolveList, ValueOrPromise} from './value-promise';
+const debug = debugFactory('loopback:context:view');
+const nextTick = promisify(process.nextTick);
+
+/**
+ * `ContextView` provides a view for a given context chain to maintain a live
+ * list of matching bindings and their resolved values within the context
+ * hierarchy.
+ *
+ * This class is the key utility to implement dynamic extensions for extension
+ * points. For example, the RestServer can react to `controller` bindings even
+ * they are added/removed/updated after the application starts.
+ *
+ */
+export class ContextView<T = unknown> implements ContextObserver {
+  protected _cachedBindings: Readonly<Binding<T>>[] | undefined;
+  protected _cachedValues: T[] | undefined;
+  private _subscription: Subscription | undefined;
+
+  constructor(
+    protected readonly context: Context,
+    public readonly filter: BindingFilter,
+  ) {}
+
+  /**
+   * Start listening events from the context
+   */
+  open() {
+    debug('Start listening on changes of context %s', this.context.name);
+    return (this._subscription = this.context.subscribe(this));
+  }
+
+  /**
+   * Stop listening events from the context
+   */
+  close() {
+    debug('Stop listening on changes of context %s', this.context.name);
+    if (!this._subscription || this._subscription.closed) return;
+    this._subscription.unsubscribe();
+    this._subscription = undefined;
+  }
+
+  /**
+   * Get the list of matched bindings. If they are not cached, it tries to find
+   * them from the context.
+   */
+  get bindings(): Readonly<Binding<T>>[] {
+    debug('Reading bindings');
+    if (this._cachedBindings == null) {
+      this._cachedBindings = this.findBindings();
+    }
+    return this._cachedBindings;
+  }
+
+  /**
+   * Find matching bindings and refresh the cache
+   */
+  protected findBindings() {
+    debug('Finding matching bindings');
+    this._cachedBindings = this.context.find(this.filter);
+    return this._cachedBindings;
+  }
+
+  /**
+   * Listen on `bind` or `unbind` and invalidate the cache
+   */
+  observe(event: ContextEventType, binding: Readonly<Binding<unknown>>) {
+    this.refresh();
+  }
+
+  /**
+   * Refresh the view by invalidating its cache
+   */
+  refresh() {
+    debug('Refreshing the view by invalidating cache');
+    this._cachedBindings = undefined;
+    this._cachedValues = undefined;
+  }
+
+  /**
+   * Resolve values for the matching bindings
+   * @param session Resolution session
+   */
+  resolve(session?: ResolutionSession): ValueOrPromise<T[]> {
+    debug('Resolving values');
+    // We don't cache values with this method as it returns `ValueOrPromise`
+    // for inject `resolve` function to allow `getSync` but `this._cachedValues`
+    // expects `T[]`
```

Comment:
Re-posting https://github.com/strongloop/loopback-next/pull/2122#discussion_r245917754:

The name `resolve` does not provide any hints that it is completely bypassing the cache. Let's find a better name please, e.g. `fetchCurrentValues`.

I understand why it may be desirable to get the values directly from the context, bypassing the value cache.

What I don't understand is why we are not updating the cache with the obtained values? Why would we want to keep outdated values in the cache when we have fresh data at our fingertips?

---

> We don't cache values with this method as it returns `ValueOrPromise`
> for inject `resolve` function to allow `getSync` but `this._cachedValues`
> expects `T[]`

What is preventing us from updating the cache after the promise is resolved? Something along the following lines:

```ts
const result = resolveList(this.bindings, b => {
  return b.getValue(this.context, ResolutionSession.fork(session));
});

if (isPromise(result)) {
  result.then(r => {this._cachedValues = r; return r;})
} else {
  this._cachedValues = result;
}
return result;
```

Reply:
I guess I use `async/await` too much. Yes, the `then` handler can update the cached values.